File: sdk/dart/dovahlink_client/lib/src/protocol/snapshot_request_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';

part 'snapshot_request_payload.g.dart';

/// Outgoing `snapshot_request` payload (`protocol/schema/README.md`'s `snapshot_request`).
/// Encode-only: the client never decodes its own `snapshot_request`.
@JsonSerializable(createFactory: false)
class SnapshotRequestPayload {
  /// The state area whose snapshot is requested.
  final String stateArea;

  /// The client's latest known revision, when available; advisory only. Omitted from the encoded
  /// object entirely (not merely `null`) when absent.
  @JsonKey(includeIfNull: false)
  final int? knownRevision;

  /// Creates a snapshot-request payload.
  const SnapshotRequestPayload({required this.stateArea, this.knownRevision});

  /// Encodes this payload as a JSON object.
  JsonMap toJson() => _$SnapshotRequestPayloadToJson(this);
}

File: integration/DovahLinkValidationClient/SnapshotRequestPayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Outgoing <c>snapshot_request</c> payload (<c>protocol/schema/README.md</c>'s
/// <c>snapshot_request</c>). Encode-only: this client never decodes its own
/// <c>snapshot_request</c>.
/// </summary>
/// <param name="StateArea">The state area whose snapshot is requested.</param>
/// <param name="KnownRevision">The client's latest known revision, when available; advisory only.
/// Omitted from the encoded object entirely (not merely <see langword="null"/>) when absent.</param>
public sealed record SnapshotRequestPayload(string StateArea, int? KnownRevision = null)
{
    /// <summary>
    /// Encodes this payload as a JSON object.
    /// </summary>
    /// <returns>The encoded <c>snapshot_request</c> payload.</returns>
    public JsonObject Encode()
    {
        var obj = new JsonObject { ["stateArea"] = StateArea };
        if (KnownRevision is not null)
        {
            obj["knownRevision"] = KnownRevision;
        }
        return obj;
    }
}
